src/main/java/fr/sncf/osrd/infra_state/SwitchPost.java

```diff
@@ -0,0 +1,196 @@
+import fr.sncf.osrd.infra.Infra;
+
+import java.util.List;
+import java.util.HashMap;
+import java.util.HashSet;
+
+public class SwitchPost {
+
+    private HashMap<String, SuccessionTable> tables;
+    private HashMap<String, Integer> currentIndex;
+    private HashMap<String, HashMap<String, Integer>> occurences;
+    private HashMap<String, HashSet<Request>> waitingList;
+    private HashMap<String, String> lastRequestedRoute;
+    private HashMap<String, String> currentTrainAllowed;
+
+    public SwitchPost() {
+        tables = null;
+    }
+
+    /**
+     * fill the fields of the SwitchPost according to the given infra and the initial trains successions tables
+     * each switch should be in the given trains successions tables list
+     * @param infra the infra
+     * @param initTables the initial train successions table
+     */
+    public void init(Infra infra, List<SuccessionTable> initTables) {
+        // build succession tables from initTables
+        tables = new HashMap<>();
+        currentIndex = new HashMap<>();
+        occurences = new HashMap<>();
+        for (var table : initTables) {
+            tables.put(table.switchID, table.clone());
+            currentIndex.put(table.switchID, 0);
+            occurences.put(table.switchID, new HashMap<String, Integer>());
+            for (var trainID : table.table) {
+                plan(table.switchID, trainID);
+            }
+        }
+
+        // build waiting list for each TVDSection
+        waitingList = new HashMap<>();
+        for (var tvdSection : infra.tvdSections.values()) {
+            waitingList.put(tvdSection.id, new HashSet<Request>());
+        }
+
+        // build last request table for train
+        lastRequestedRoute = new HashMap<>();
+        currentTrainAllowed = new HashMap<>();
+    }
+
+    /**
+     * check if the switch switchID is reserved for the train trainID
+     * @param switchID the switch id
+     * @param trainID the train id
+     * @return true iff the given switch is reserved for the given train
+     */
+    public boolean isCurrentAllowed(String switchID, String trainID) {
+        return currentTrainAllowed.containsKey(trainID) && currentTrainAllowed.get(switchID).equals(trainID);
+    }
+
+    private boolean isPlanned(String switchID, String trainID) {
+        assert tables.containsKey(switchID);
+        return occurences.get(switchID).containsKey(trainID) && occurences.get(switchID).get(trainID) > 0;
+    }
+
+    private boolean isNext(String switchID, String trainID) {
+        assert tables.containsKey(switchID);
+        var index = currentIndex.get(switchID);
+        return tables.get(switchID).table.get(index).equals(trainID);
+    }
+
+    private void plan(String switchID, String trainID) {
+        assert tables.containsKey(switchID);
+        tables.get(switchID).table.add(trainID);
+        var count = occurences.get(switchID).containsKey(trainID) ? occurences.get(switchID).get(trainID) : 0;
```

Comment:
Most of the functions here can be changed like that:

```suggestion
        var count = occurences.get(switchID).getOrDefault(trainID, 0);
```